File: app/src/main/java/com/intkhabahmed/smartnotes/database/NoteRepository.java
package com.intkhabahmed.smartnotes.database;

import android.arch.lifecycle.LiveData;
import android.arch.paging.DataSource;
import android.text.TextUtils;

import com.intkhabahmed.smartnotes.models.Note;
import com.intkhabahmed.smartnotes.utils.AppConstants;
import com.intkhabahmed.smartnotes.utils.AppExecutors;
import com.intkhabahmed.smartnotes.utils.Global;

public class NoteRepository {
    private static NoteRepository sInstance;
    private static final Object LOCK = new Object();

    public static NoteRepository getInstance() {
        if (sInstance == null) {
            synchronized (LOCK) {
                sInstance = new NoteRepository();
            }
        }
        return sInstance;
    }

    public DataSource.Factory<Integer, Note> getNotesByTypeAndAvailability(String type, int trashed) {
        if (TextUtils.isEmpty(type)) {
            return Global.getDbInstance().notesDao().getNotesByAvailability(trashed);
        }
        if (Global.getSortCriteria().contains(AppConstants.ASC)) {
            return Global.getDbInstance().notesDao().getNotesByTypeAndAvailabilityInAscendingOrder(type, trashed,
                    Global.getSortCriteria());
        }
        return Global.getDbInstance().notesDao().getNotesByTypeAndAvailabilityInDescendingOrder(type, trashed,
                Global.getSortCriteria());
    }

    public DataSource.Factory<Integer, Note> getNotesByTitleAndAvailability(String title, int trashed) {
        return Global.getDbInstance().notesDao().getNotesByTitleAndAvailability(title, trashed);
    }

    public long insertNote(Note note) {
        return Global.getDbInstance().notesDao().insertNote(note);
    }

    public int updateNote(Note note) {
        return Global.getDbInstance().notesDao().updateNote(note);
    }

    public void deleteNote(Note note) {
        Global.getDbInstance().notesDao().deleteNote(note);
    }

    public void moveNoteToTrash(final Note note) {
        note.setTrashed(1);
        AppExecutors.getInstance().diskIO().execute(new Runnable() {
            @Override
            public void run() {
                Global.getDbInstance().notesDao().updateNote(note);
            }
        });
    }

    public void recoverNoteFromTrash(final Note note) {
        note.setTrashed(0);
        AppExecutors.getInstance().diskIO().execute(new Runnable() {
            @Override
            public void run() {
                Global.getDbInstance().notesDao().updateNote(note);
            }
        });
    }

    public LiveData<Note> getNoteById(int noteId) {
        return Global.getDbInstance().notesDao().getNoteById(noteId);
    }

    public void emptyTrash() {
        AppExecutors.getInstance().diskIO().execute(new Runnable() {
            @Override
            public void run() {
                Global.getDbInstance().notesDao().deleteAll();
            }
        });
    }
}

File: app/src/main/java/com/intkhabahmed/smartnotes/database/NoteRepository.kt
package com.intkhabahmed.smartnotes.database

import android.text.TextUtils
import androidx.lifecycle.LiveData
import androidx.paging.DataSource
import com.intkhabahmed.smartnotes.models.Note
import com.intkhabahmed.smartnotes.utils.AppConstants
import com.intkhabahmed.smartnotes.utils.AppExecutors
import com.intkhabahmed.smartnotes.utils.Global

class NoteRepository {
    fun getNotesByTypeAndAvailability(type: String?, trashed: Int): DataSource.Factory<Int, Note> {
        if (TextUtils.isEmpty(type)) {
            return Global.getDbInstance().notesDao().getNotesByAvailability(trashed)
        }
        return if (Global.getSortCriteria().contains(AppConstants.ASC)) {
            Global.getDbInstance().notesDao().getNotesByTypeAndAvailabilityInAscendingOrder(type, trashed,
                    Global.getSortCriteria())
        } else Global.getDbInstance().notesDao().getNotesByTypeAndAvailabilityInDescendingOrder(type, trashed,
                Global.getSortCriteria())
    }

    fun getNotesByTitleAndAvailability(title: String?, trashed: Int): DataSource.Factory<Int, Note> {
        return Global.getDbInstance().notesDao().getNotesByTitleAndAvailability(title, trashed)
    }

    fun insertNote(note: Note?): Long? {
        return Global.getDbInstance().notesDao().insertNote(note)
    }

    fun updateNote(note: Note?): Int? {
        return Global.getDbInstance().notesDao().updateNote(note)
    }

    fun deleteNote(note: Note?) {
        Global.getDbInstance().notesDao().deleteNote(note)
    }

    fun moveNoteToTrash(note: Note) {
        note.trashed = 1
        AppExecutors.getInstance().diskIO().execute { Global.getDbInstance().notesDao().updateNote(note) }
    }

    fun recoverNoteFromTrash(note: Note) {
        note.trashed = 0
        AppExecutors.getInstance().diskIO().execute { Global.getDbInstance().notesDao().updateNote(note) }
    }

    fun getNoteById(noteId: Int): LiveData<Note> {
        return Global.getDbInstance().notesDao().getNoteById(noteId)
    }

    fun emptyTrash() {
        AppExecutors.getInstance().diskIO().execute { Global.getDbInstance().notesDao().deleteAll() }
    }

    companion object {
        private var sInstance: NoteRepository? = null
        private val LOCK = Any()

        @JvmStatic
        val instance: NoteRepository?
            get() {
                if (sInstance == null) {
                    synchronized(LOCK) { sInstance = NoteRepository() }
                }
                return sInstance
            }
    }
}